ported the joint_trajectory_controller query_state service to ROS2

saikishor opened the pull request:
Hello!

There is a missing query_state from the ROS1 `joint_trajectory_controllers` : https://github.com/ros-controls/ros_controllers/blob/noetic-devel/joint_trajectory_controller/include/joint_trajectory_controller/joint_trajectory_controller_impl.h#L662-L708. I've ported this service to the ROS2 version of `joint_trajectory_controllers`

destogl reviewed (commented):
Thanks for your contribution :)

I understand that this is available in ROS 1, but what I am uncertain if this is really needed, since the same data are available on the `state` topic.

Can you describe a bit more purpose of this service in your application?

saikishor commented:
Hello @destogl!

> I understand that this is available in ROS 1, but what I am uncertain if this is really needed, since the same data are available on the `state` topic.
> 
> Can you describe a bit more purpose of this service in your application?


Thanks for the imminent response. The idea of this query state is that you can be able to query the future state of the trajectory. For instance, you can pass a timestamp 1 second in the future, and get the state it is going to be.

I can be useful for different usecases, for instance, after sending the trajectory, I could utilise it to check collisions in future joint references and also can use it as a prediction so if my next trajectory is around that point, I can avoid sending it etc.

I hope this answers your question. :)

bmagyar commented:
As we discussed on Slack, let's wait until there's a tiny refactor & a basic test on the PR

saikishor commented:
> As we discussed on Slack, let's wait until there's a tiny refactor & a basic test on the PR

@bmagyar I've reiterated through the MR and modified my initial approach and simplified it much further as the earlier instance seems to have duplicated checks. Please check and let me know what you think.

saikishor commented:
@bmagyar The failing tests seems to be of admittance control, and binary build seems to have issue with the rosdep. Ofcourse, these changes break ABI🤐

saikishor commented on joint_trajectory_controller/src/joint_trajectory_controller.cpp:
```diff
@@ -462,6 +462,53 @@ bool JointTrajectoryController::read_state_from_command_interfaces(JointTrajecto
+  std::shared_ptr<control_msgs::srv::QueryTrajectoryState::Response> response)
+{
+  const auto logger = get_node()->get_logger();
+  const auto active_goal = *rt_active_goal_.readFromRT();
+  response->name = params_.joints;
+  trajectory_msgs::msg::JointTrajectoryPoint state_requested = state_current_;
```
Agreed!
Thank you for the feedback @bmagyar

saikishor replied:
Updated as requested. :)

bmagyar replied:
Grand!

saikishor commented on joint_trajectory_controller/src/joint_trajectory_controller.cpp:
```diff
@@ -462,6 +462,53 @@ bool JointTrajectoryController::read_state_from_command_interfaces(JointTrajecto
+  // Preconditions
+  if (get_state().id() != lifecycle_msgs::msg::State::PRIMARY_STATE_ACTIVE)
+  {
+    RCLCPP_ERROR(logger, "Can't sample trajectory. Controller is not running.");
```
Done

bmagyar replied:
thx

bmagyar commented:
@mergifyio backport humble
